src/workerd/io/actor-sqlite.c++

```diff
@@ -396,22 +406,38 @@ kj::Promise<void> ActorSqlite::commitImpl(ActorSqlite::PrecommitAlarmState preco
   fulfiller->fulfill();
 
   // If the db state is now later than the in-flight scheduled alarms, issue a request to update
-  // it to match the db state.  We don't need to hold open the output gate, so we add the
-  // scheduling request to commitTasks.
+  // it to match the db state. We chain these requests to ensure they execute in order and
+  // prevent races at the alarm manager.
   if (willFireEarlier(alarmScheduledNoLaterThan, alarmStateForCommit)) {
-    alarmLaterTasks.add(requestScheduledAlarm(alarmStateForCommit));
+    // We need to cancel our pending alarm chain cleanup task before we schedule a new "move later"
+    // alarm. Otherwise, we could unintentionally cancel our in-flight alarmLaterChain tasks.
+    alarmChainCleanup = kj::none;
+    auto updatePromise = [this, alarmStateForCommit]() -> kj::Promise<void> {
```

Comment:
This is a coroutine lambda with captures which is (later) passed to `.then()`. This requires `kj::coCapture()`.